internal/ai/conversation.go

```diff
@@ -157,16 +172,190 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+	// to maintain context while summarizing the middle portion
+	preserveFirstMessages := 1
+	preserveLastMessages := 2
+	
+	// Ensure we have something to summarize
+	if len(cc.Messages) <= preserveFirstMessages + preserveLastMessages {
+		return nil
+	}
+
+	// Extract messages to summarize (middle portion)
+	startIdx := preserveFirstMessages
+	endIdx := len(cc.Messages) - preserveLastMessages
+	messagesToSummarize := cc.Messages[startIdx:endIdx]
+	
+	if len(messagesToSummarize) == 0 {
+		return nil
+	}
+
+	// Generate summary using AI
+	summary, err := cc.generateConversationSummary(ctx, messagesToSummarize)
+	if err != nil {
+		return fmt.Errorf("failed to generate conversation summary: %w", err)
+	}
+
+	// Create a summary message
+	summaryMessage := conversationTurn{
+		UserMessage: anthropic.NewUserMessage(anthropic.NewTextBlock(summary)),
+		// No response - this is just context
+	}
+
+	// Reconstruct the conversation with preserved messages + summary + recent messages
+	newMessages := []conversationTurn{}
+	
+	// Add preserved first messages
+	newMessages = append(newMessages, cc.Messages[:preserveFirstMessages]...)
+	
+	// Add summary message
+	newMessages = append(newMessages, summaryMessage)
+	
+	// Add preserved last messages
+	newMessages = append(newMessages, cc.Messages[endIdx:]...)
+
+	// Update the conversation
+	cc.Messages = newMessages
+	
+	// Reset token counters since we've condensed the conversation
+	cc.totalInputTokens = 0
+	cc.totalCacheReadTokens = 0
+
+	log.Printf("Conversation summarized: %d messages -> %d messages", 
+		len(messagesToSummarize) + preserveFirstMessages + preserveLastMessages, 
+		len(cc.Messages))
+
+	return nil
+}
+
+// generateConversationSummary creates a summary of the given conversation turns using AI
+func (cc *Conversation) generateConversationSummary(ctx context.Context, messages []conversationTurn) (string, error) {
+	// Create a separate conversation for summarization to avoid infinite recursion
+	summaryConv := &Conversation{
+		client:       cc.client,
+		model:        cc.model,
+		systemPrompt: "You are a helpful assistant that creates concise summaries of coding conversations.",
+		maxTokens:    4000, // Limit summary length
+		tokenLimit:   0, // No token limit for summary conversation
+	}
+
+	// Build a summary request
+	var conversationText strings.Builder
+	conversationText.WriteString("Please create a concise summary of this coding conversation. Focus on:\n")
+	conversationText.WriteString("1. Key decisions and changes made\n")
+	conversationText.WriteString("2. Current state of the codebase\n")
+	conversationText.WriteString("3. Any important context for continuing the work\n")
+	conversationText.WriteString("4. Tools used and their outcomes\n\n")
+	conversationText.WriteString("Conversation to summarize:\n\n")
+
+	// Convert messages to readable format
+	for i, turn := range messages {
+		conversationText.WriteString(fmt.Sprintf("=== Turn %d ===\n", i+1))
```

Comment:
We should send the summary request as part of the original conversation so that the history is automatically included